Update to probe and unprotect sequence

HenryNguyen01 opened the pull request:
Updating the probe and unprotect sequence to have an understanding of the different flash versions between MSPM0 devices.

nmenon committed:
flash/nor: Add basic support for TI's MSPM0L/G family

Add basic flashing support for Texas Instruments MSPM0L, C and G family of
Cortex-M0 based micro-controllers.

This initial basic flashing support allows for controlling protection,
erase, write and read of non-main flash region.

This has been tested with:
* valgrind --leak-check=full --show-leak-kinds=all \
	--track-origins=yes --verbose
* Ubuntu clang version 19.0.0
 (++20240627042246+c791d86eab13-1~exp1~20240627042407.1770)

Valgrind-clean, no new Clang analyzer warnings have been introduced.

Change-Id: I29b8055ea6da9c38c5b7b91bea1ec7581c5bc8ff
Co-developed-by: Henry Nguyen <h-nguyen8@ti.com>
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>
Signed-off-by: Nishanth Menon <nm@ti.com>

nmenon committed:
tcl/target: Add support for TI MSPM0

Add basic support for Texas Instruments MSPM0L, C and G family of
Cortex-M0 based micro-controllers.

Change-Id: If2b5b1eca001f74d501ede67ec621c7497548a85
Co-developed-by: Henry Nguyen <h-nguyen8@ti.com>
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>
Signed-off-by: Nishanth Menon <nm@ti.com>

nmenon committed:
tcl/board: Add MSPM0 Launchpad support

Add basic connection details for TI's MSPM0 Launchpad series of
evaluation kits:
https://www.ti.com/tool/LP-MSPM0L1306
https://www.ti.com/tool/LP-MSPM0C1104
https://www.ti.com/tool/LP-MSPM0G3507

Change-Id: I33499f2d5fef846185ff6c330f9bfd0251117eb6
Co-developed-by: Henry Nguyen <h-nguyen8@ti.com>
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>
Signed-off-by: Nishanth Menon <nm@ti.com>

nmenon committed:
[NOT FOR UPSTREAM] Add a couple of test scripts

Lets add a few test scripts for us to easily program the device.

Change-Id: Id3ae7e362433e00ff8b0239698d46a0f1463be21
Signed-off-by: Nishanth Menon <nm@ti.com>

Commit:
ifix up for https://review.openocd.org/c/openocd/+/8384/comment/82b871fe_ce7a68dd/

Squash it

Change-Id: I52e5375fcdacd65303c052588c97eddb435a8e2b
Signed-off-by: Nishanth Menon <nmenon@ula0133052-001.dhcp.ti.com>

Commit:
fix up https://review.openocd.org/c/openocd/+/8384/comment/5efb7ece_f3b0ca61/

https://review.openocd.org/c/openocd/+/8384/comment/a6cc9617_dda0c2fc/

Change-Id: Ib4508248a12e866ebdb5821dad28fcc194493e4e
Signed-off-by: Nishanth Menon <nmenon@ula0133052-001.dhcp.ti.com>

Commit:
fix https://review.openocd.org/c/openocd/+/8384/comment/406e6d33_15247608/

Change-Id: Ib47066cbafe7a4bb1e69d1cdbdde721de82025b3
Signed-off-by: Nishanth Menon <nmenon@ula0133052-001.dhcp.ti.com>

Commit:
https://review.openocd.org/c/openocd/+/8384/comment/b035f2ec_ce46a82e/

Change-Id: Ibcdd9b7b3f38c80371af2ec34f84bd8da90892a2
Signed-off-by: Nishanth Menon <nmenon@ula0133052-001.dhcp.ti.com>

Commit:
https://review.openocd.org/c/openocd/+/8384/comment/407d381c_401299c5/

Change-Id: I03314ed0bce81752d0e093bccb0a6e4359b4752f
Signed-off-by: Nishanth Menon <nmenon@ula0133052-001.dhcp.ti.com>

Commit:
https://review.openocd.org/c/openocd/+/8384/comment/ba733e34_4d4ccf0b/

Change-Id: I5b79407a78b6b476a2da630ca509a364d170f87f
Signed-off-by: Nishanth Menon <nmenon@ula0133052-001.dhcp.ti.com>

HNguyen00 committed:
flash/nor: Update mspm0_probe() and mspm0_read_part_info() sequence

Update mspm0_probe() sequence to have an understanding of
the different flashctl feature versions and allowing for
mspm0_read_part_info() sequence to continue in the scenario that
device and part info cannot be found but the ALWAYS_1 and manufacturer
ID can be seen.

Change-Id: Idad42755a72e3bbcc7b825000fa76e8dd3244ed5
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor: Update flash command configuration

Update flash command configuration to utilize mspm0_fctl_cfg_command
to condense flash configuration down to one line.

Change-Id: I4226761a23e7e3103b03feacb28e4e8d8b468bf1
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor: Add sector and bank check

Before unprotecting any sector of memory we will now check to make
sure that the bank is known and the address falls under the proper
bank.

Change-Id: Ib937154fbb104aa101644b5fc58996a2492f3194
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor: Adding understanding of flash version B to get_sector() function

Previously the implementation of the get_sector() function could
only support flash version A and MAIN/NONMAIN banks. Update
was done so an understanding of the flash version B and DATA
bank is present within the get_sector() function.

Change-Id: I0e145fb8930df293761348758ef8ef0f40a4f5e7
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

nmenon commented on src/flash/nor/mspm0.c:
change your authorship and signoff

HNguyen00 committed:
flash/nor: Fixing formatting issues

Updating indent to be properly aligned within probe_status switch-case statement.
Adding break to unprotect_sector switch-case default statement.
Changing get_sector() switch-case to have MAIN and DATA follow the same
sequence as they both are treated the same in term of protection.

Change-Id: I4233671f3d1d9ac105a8083e69d5449e10f02463
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor: Condensing erase configuration

Further condensing the flashctl configuration to perform
an erase down to one single line within mspm0_erase

Change-Id: Ibf29adb4082d5a72a529676e3201be5e27ef4fe8
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

nmenon commented on src/flash/nor/mspm0.c:
```diff
@@ -597,10 +598,6 @@ static int mspm0_fctl_get_sector_reg(struct flash_bank *bank, unsigned int addr,
 		break;
-	case MSPM0_FLASH_BASE_DATA:
-		*sector_mask = BIT((sector_in_bank / 8) % 32);
-		*reg = FCTL_REG_CMDWEPROTB;
```
explain this change - this does'nt look like formatting fixup?

HenryNguyen01 replied:
The DATA case will perform the same exact action as MAIN because all devices with DATA bank will use only CMDWEPROTB for protection, which is why the MAIN and DATA are thrown under the same case. This technically is not a formatting change. It's more of condensing.

HNguyen00 committed:
flash/nor: Updating NONMAIN size definition

Within mspm0_probe() size definition of NONMAIN memory only defines
half of its actual size. Updating size to properly reflect NONMAIN's
size of 1kB (1024 bytes) rather than 512 bytes.

Change-Id: Ibe4cd805f173ca60a27d596e60a7294e804b71bb
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor: Updating mspm0_erase() and mspm0_write() to support all banks

Updating mspm0_erase to support sector erase on all sector of memory
within MAIN, NONMAIN, and DATA memory. Updating write function to
properly include memory offset depending on what bank is being
written to.

Change-Id: I28dcd0553695d6b82a00c2faecc8e8ff9d704698
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor/mspm0: Updating mspm0_fctl_sector_erase() comments

Add explanation as to why sector unprotection must be done before
a sector erase operation can performed.

Change-Id: I614f2bc439d51425b21c67b86eb3ba4cf25d1f54
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor/mspm0: Update mspm0_write() to use addr instead of offset

Update mspm0_write() to use addr instead of offset to improve on
code clarity when configuring flashctl to perform operations on
an address.

Change-Id: Iced79fe98fb21e8d44124d43b45baaa8e3532174
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor/mspm0: Fix to bank swap bug in get_sector_reg()

Add check to only perform a bank swap check if the bank being operated
on is MAIN bank. If the bank being operated on is NONMAIN or DATA,
then the check is ignored.

Change-Id: I384fd521540ea3c998ebe5f276037009323ed64c
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor/mspm0: Fix to comment formatting issues

Certain comments through the code did not have proper space after
the text. Update is to have all one line comments to follow the
same formatting.

Change-Id: Ia0b1a1aeadb1cd6c6fefca4027c665bd85e96a38
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HenryNguyen01 committed:
flash/nor/mspm0: Fix to calloc usage

Calloc usage had the entries flipped where the number
of elements allocated was actually the size of elemnt
and vice-versa.

Change-Id: Ieb39ba35646c4128597a41f111a2aafba7f842c4
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

HNguyen00 committed:
flash/nor/mspm0: Removing unnecessary typecast

Previously typecast were performed when passing in addresses for
flash operation. They however, are not required so they are
removed.

Change-Id: Id117b0de3a8c97062e74cc3f1ad32304bf968652
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

nmenon commented:
Squashed
